### PR TITLE
Add Eigen as an external project

### LINKED ISSUE #18540: Add Eigen as an External Project
Add Eigen as an external project to Mantid, for use in Instrument 2.0. Using an external project will ensure we get the same version on all platforms and no extra dependencies will be required.

~~Wrapping `Matrix3d` and `Vector3d` with `boost:python` should also be investigated as part of this ticket.~~

### PR DESCRIPTION
**Pull request by ianbush**
Add Eigen as an external dependency.

**To test:**

Checkout the branch and re-run `cmake` however you normally would. Check that there are no errors.

Try running `cmake` in a clean directory and check there are no errors.

You should also checkout out commit 9de61cf and try building the `Beamline` target. You can see the build history to see that this test passed on the build servers (`Eigen` will soon be used in `Beamline`).

Check no extra targets are created, e.g. `ninja -t targets` or check the list in VS.

Fixes #18540.

*Internal change, does not need to be in the release notes - but developer documentation should be updated once this is merged.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by LamarMoore** [comment]
Works fine for me on Windows

File: Framework/Beamline/inc/MantidBeamline/EigenTest.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#ifndef MANTID_EIGENTEST_H
```
Comment (SimonHeybrock) [inline Framework/Beamline/inc/MantidBeamline/EigenTest.h]: Can we remove the header and simply put everything in the test? That would avoid polluting the library and no one will mind keeping the test around.

File: buildconfig/CMake/Eigen.cmake [hunk 2]
```diff
@@ -0,0 +1,25 @@
+include(ExternalProject)
+
+# Use version 3.2.10 of Eigen
+set(eigen_version "3.2.10")
```
Comment (SimonHeybrock) [inline buildconfig/CMake/Eigen.cmake]: Owen suggest to add a comment here why you went back to the old version (compiler warnings in the latest builds).

**Comment by SimonHeybrock** [comment]
Tested on Ubuntu, cmake in clean dir works without errors, and so does rerunning cmake.

File: Framework/Beamline/inc/MantidBeamline/EigenTest.h (hunk 1 above)
Comment (ianbush) [inline Framework/Beamline/inc/MantidBeamline/EigenTest.h]: I was planning to remove `EigenTest.h` and `EigenTest.cpp` completely before removing 'In Progress'. It was just to test the build servers can actually build something using `Eigen`.

Comment (SimonHeybrock) [inline Framework/Beamline/inc/MantidBeamline/EigenTest.h]: Also fine :+1:

**Comment by SimonHeybrock** [comment]
Owen also tested this on OSX, works -> :shipit:

**Pull request merged**